src/box/txn_limbo.c

```diff
@@ -393,44 +466,116 @@ synchro_request_write(const struct synchro_request *req)
+	if (limbo->owner_id == instance_id) {
+		assert(lsn > limbo->confirmed_lsn);
+		assert(!limbo->is_in_rollback);
+		limbo->confirmed_lsn = lsn;
+	}
 
-	struct synchro_request req = {
-		.type = type,
-		.replica_id = limbo->owner_id,
+	synchro_request_write(&(struct synchro_request) {
+		.type = IPROTO_RAFT_CONFIRM,
+		.replica_id = instance_id,
 		.lsn = lsn,
-		.term = term,
-		.confirmed_vclock = vclock,
-	};
-	synchro_request_write(&req);
+	});
+}
+
+static void
+txn_limbo_log_status(struct txn_limbo *limbo)
+{
+	say_info("PROMOTE: limbo owner_id: %u, "
+		 "lsn: %" PRIi64 ", term: %" PRIu64,
+		 limbo->owner_id, limbo->confirmed_lsn,
+		 limbo->promote_greatest_term);
 }
 
 /**
- * Write a confirmation entry to WAL. After it's written all the
- * transactions waiting for confirmation may be finished.
+ * Confirm a queue (history) of PROMOTE requests on the leader's behalf.
+ * All the matching requests will be applied in order.
  */
 static void
-txn_limbo_write_confirm(struct txn_limbo *limbo, int64_t lsn)
+txn_limbo_confirm_promote(struct txn_limbo *limbo,
+			  uint32_t origin_id, int64_t lsn)
 {
-	assert(lsn > limbo->confirmed_lsn);
-	assert(!limbo->is_in_rollback);
-	limbo->confirmed_lsn = lsn;
-	vclock_follow(&limbo->confirmed_vclock, limbo->owner_id, lsn);
-	txn_limbo_write_synchro(limbo, IPROTO_RAFT_CONFIRM, lsn, 0, NULL);
+	assert(latch_is_locked(&limbo->promote_latch));
+
+	say_info("PROMOTE: read %s for lsn %" PRIi64 " from %u",
+		 iproto_type_name(IPROTO_RAFT_CONFIRM), lsn, origin_id);
+
+	struct rlist *queue = &limbo->pending_promotes;
+
+	uint64_t prev_term = 0;
+	struct pending_promote *item, *tmp, *confirmed = NULL;
+	rlist_foreach_entry_safe_reverse(item, queue, link, tmp) {
+		/*
+		 * Find the latest matching request; Then, drop all
+		 * requests which can no longer belong to the history
+		 * (i.e. unreachable from the confirmed one).
+		 */
+		if (confirmed != NULL) {
+			if (item->req.term == prev_term) {
+				prev_term = item->req.prev_term;
+				continue;
+			}
+
+			say_info("PROMOTE: drop obsolete %s",
+				 synchro_request_to_string(&item->req));
+
+			rlist_del_entry(item, link);
+			free(item);
+
+		} else if (item->req.origin_id == origin_id &&
+			   item->req.self_lsn <= lsn) {
+			say_info("PROMOTE: confirm %s",
+				 synchro_request_to_string(&item->req));
+
+			confirmed = item;
+			prev_term = item->req.prev_term;
+		}
+	}
+
+	/*
+	 * If we couldn't find the entry, it means that the CONFIRM is
+	 * obsolete. This is fine, because it implies that the PROMOTE
+	 * has already been applied as a part of a longer history.
+	 */
+	if (confirmed != NULL) {
+		/* Process all confirmed PROMOTE requests in order. */
+		rlist_foreach_entry_safe(item, queue, link, tmp) {
+			struct synchro_request *req = &item->req;
+			txn_limbo_apply_promote(limbo, req, req->origin_id);
+			rlist_del_entry(item, link);
+			free(item);
+
+			if (item == confirmed)
+				break;
+		}
```

Comment:
Can't find where do you sort the promotes. Are they sorted by term? The only place where promotes are being added to the queue is `txn_limbo_make_promote()`. Can it happen, than node-1 makes a PROMOTE, then hangs, then node-2 bumps term, makes PROMOTE too. Then node-1 unhangs and node-2 receives an older PROMOTE from node-1. Looks like it should be placed in the beginning of the queue, but from the code it would be placed in the end, violating the order. I am 99% sure I've missed something. Could you help?